### DISCUSSION TITLE
Custom deck for two TCS34725 RGB sensors and TCA9548A I2C MUX

### DISCUSSION THREAD

**Discussion by GoLut** [bitcraze/discussions#462] [Q&A] [answered]
Dear Crazyflie DEVs,

I am currently working on a project that involves integrating 2x: [TCS34725](https://www.adafruit.com/product/1334) RGB color sensor and a [TCA9548A](https://learn.adafruit.com/adafruit-tca9548a-1-to-8-i2c-multiplexer-breakout) 1-to-8 I2C multiplexer on the Crazyflie. I have done some research and the best approach seems to be developing my own deck using either the [prototyping](https://www.bitcraze.io/products/prototyping-deck/) deck or the [breakout](https://www.bitcraze.io/products/breakout-deck/) deck. Both of which I have in my possession.

**Q1**: Is this the correct approach? Do the DEVs (or anyone else reading this) recommend a different approach?

The RGB sensors have the following interfaces that I need to use:

- I2C for communication to and from the sensor. 	
- Interrupt pin that is pulled LOW when new data is available. 

Currently, I have a C-based implementation of the whole system working on an Arduino DUE. 
Whenever the Interrupt is detected the system runs a small function. And afterward, the sensor data is read using I2C.

The next step that I have to do is to migrate my code to the Crazyflie.
I did some research online and found good examples of I2C sensors being implemented on the Crazyflie: (A reference to other users)

1. pca9685 sensor driver. [Link](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/drivers/src/pca9685.c)
2. I2C EEPROM: [Link](https://forum.bitcraze.io/viewtopic.php?t=3149)
3. BH-1750 Color Sensor: [Link](https://www.bitcraze.io/2020/07/intro-to-autonomous-robotics-with-the-crazyflie/)
4. Hall effect sensor: [Link](https://github.com/bitcraze/discussions/discussions/306)

With the above information, I'm sure I can figure it out for my own system. Using the functions provided in the “i2cdev.c” source file. 

In addition, I have found some information on how to import external driver files here: [Link](https://github.com/bitcraze/discussions/discussions/414). 

The one part I'm still missing is how to use a form of GPIO interrupt. E.g. Whenever the pin is pulled LOW some code is run by the system. 
Maybe FreeRTOS has some functionality that I can use? (Not as familiar yet with all the functionalities).

**Q2**: Could the DEVs help here and point me in the right direction? 

I did however find some workarounds achieving a similar result:
- FreeRTOS Timer based [Link](https://forum.bitcraze.io/viewtopic.php?t=4115)
- PWM signal readout (Sadly no code was provided) [Link](https://forum.bitcraze.io/viewtopic.php?t=1783)

Thank you in advance for answering my questions.

**Comment by tobbeanton** [comment, marked as answer]
Q1: If the sensors are easy enough to solder without making a PCB design the prototyping deck should do the job.

Q2: There's no deck interface to simply use pin interrupts but you can use the system functions. First find out what pin you would like to use. If e.g. you choose IO1, this is mapped to PB8. EXTI is the STM32 pin interrupt system which you could read up on but in short PB8 is mapped to EXTI interrupt 8. The system already has [callback functions](https://github.com/bitcraze/crazyflie-firmware/blob/a4d91f9066a6426c9c4ccee7a025fdec767e5d73/src/drivers/src/exti.c#L190) for the EXTI interrupts so you have to add this to you deck driver file and also enable it. Use e.g. the [uart_syslink.c](https://github.com/bitcraze/crazyflie-firmware/blob/1bf766584db97b6d27c4d3fc17e9ab6f7e3b1180/src/drivers/src/uart_syslink.c#L620) as an example. Enabling of the interrupt is done [here](https://github.com/bitcraze/crazyflie-firmware/blob/1bf766584db97b6d27c4d3fc17e9ab6f7e3b1180/src/drivers/src/uart_syslink.c#L229). Hope this gets you started!

**Reply by GoLut** [reply]
Thank you for the fast reply!
I think with the information provided I can come to a solution. :)
